src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -167,154 +158,88 @@
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
+
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
+
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
+
+  (mutils/multitime
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
+   (f))
+
+  (mark-both-metrics! command version :processed))
+
+(defn message-handler-with-retries
+  "This function processing of the message, retrying messages that
+  fail and discarding messages that have fatal errors or have exceeded
+  their maximum allowed attempts. `delay-message-fn` and
+  `discard-message-fn` are both functions of two arguments, a
+  `message` and an `exception`. `process-message-fn` is a function
+  that accepts a message as it's argument"
+  [delay-message-fn discard-message-fn process-message-fn]
   (fn [msg]
     (try+
 
-     (mark! (global-metric :seen))
-     (time! (global-metric :processing-time)
-            (f msg))
-     (mark! (global-metric :processed))
-
-     (catch fatal? {:keys [cause]}
-       (on-fatal msg cause)
-       (mark! (global-metric :fatal)))
+     (let [{:keys [command version annotations] :as parsed-result} (parse-command msg)
+           retries (count (:attempts annotations))
+           id (:id annotations)]
 
-     (catch Throwable exception
-       (on-retry msg exception)
-       (mark! (global-metric :retried))))))
+       (try+
 
-(defn wrap-with-command-parser
-  "Wrap a message processor `f` such that all messages passed to `f` are
-  well-formed commands. If a message cannot be parsed, the `on-fatal` hook is
-  invoked, and `f` is ignored."
-  [f on-parse-error]
-  (fn [msg]
-    (let [parse-result (try-parse-command msg)]
-      (if (instance? Throwable parse-result)
-        (do
+        (call-with-command-metrics command version retries
+          #(process-message-fn parsed-result))
+        
+        (catch fatal? _
           (mark! (global-metric :fatal))
-          (on-parse-error msg parse-result))
-        (f parse-result)))))
-
-(defn wrap-with-discard
-  "Wrap a message processor `f` such that incoming commands with a
-  retry count exceeding `max-retries` are discarded.
-
-  This assumes that all incoming messages are well-formed command
-  objects, such as those produced by the `wrap-with-command-parser`
-  middleware."
-  [f on-discard max-retries]
-  (fn [{:keys [command version annotations] :as msg}]
-    (let [retries (count (:attempts annotations))]
-      (create-metrics-for-command! command version)
-      (mark! (cmd-metric command version :seen))
-      (update! (global-metric :retry-counts) retries)
-      (update! (cmd-metric command version :retry-counts) retries)
-
-      (when (>= retries max-retries)
```

Comment:
There was a bug in the old code. After the 15th attempt to process a message, we schedule a retry (maybe a day later) but when that message is processed again it gets discarded. As a result, we delay/enqueue the 16th attempt to later discard it without processing it.